### PR TITLE
Internal documentation

### LINKED ISSUE #4: Internal Documentation
macros are hard

### PR DESCRIPTION
**Pull request by daxpedda**
This adds internal documentation, I hope you like my style!
I also managed to remove proc-macro-error and left some TODOs where necessary.
Some code shuffling also happened.

Fixes #4.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,49 +1,69 @@
...
 use syn::parse::{Parse, ParseStream};
 use syn::punctuated::Punctuated;
 use syn::spanned::Spanned;
 use syn::{
-    parse_macro_input, Data, DeriveInput, Error, Fields, FieldsNamed, FieldsUnnamed, Token,
-    TraitBound, Type,
+    Data, DeriveInput, Error, Fields, FieldsNamed, FieldsUnnamed, Result, Token, TraitBound, Type,
 };
 
+/// Holds parsed [bounds](TraitBound) and [traits][Traits].
```
Comment (ModProg) [inline src/lib.rs:18]: This looks strange, shouldn't both links have the same syntax?

File: src/lib.rs [hunk 2]
```diff
@@ -1,49 +1,69 @@
...
 struct DeriveWhere {
+    /// Parsed [bounds](TraitBound).
     bounds: Vec<TraitBound>,
+    /// Parsed [traits][Traits].
```
Comment (ModProg) [inline src/lib.rs:22]: Same here, I think both work just interested why

File: src/lib.rs [hunk 3]
```diff
@@ -447,51 +491,66 @@ impl Traits {
...
+            // If there is no `where` clause, make one.
+            let where_clause = where_clause.get_or_insert(quote! { where });
 
             for bound in &derive_where.bounds {
-                bounds.extend(quote! {
+                where_clause.extend(quote! {
                     #bound: #trait_,
                 })
             }
+        }
```
Comment (ModProg) [inline src/lib.rs:530]: 1. Do we need to support empty bounds?
2. I think the DeriveWhere parser fails on empty bounds or traits.

**Review by ModProg — CHANGES_REQUESTED** [review]
https://github.com/ModProg/derive-where/blob/5712ae08e5bf06fec188e428685fbe93c1dd857f/src/lib.rs#L166
Maybe document skip, the rest is very self-explanatory 

https://github.com/ModProg/derive-where/blob/5712ae08e5bf06fec188e428685fbe93c1dd857f/src/lib.rs#L293
Maybe document pattern

File: src/lib.rs (hunk 1 above)
Comment (daxpedda) [reply src/lib.rs:18]: This is a typo.

File: src/lib.rs (hunk 2 above)
Comment (daxpedda) [reply src/lib.rs:22]: I copied the typo from above and didn't notice, I would also like to know why this works 😄.

File: src/lib.rs (hunk 3 above)
Comment (daxpedda) [reply src/lib.rs:530]: 1. We do have to support empty bounds:
```rust
struct Test<T>(PhantomData<T>);

impl<T> Clone for Test<T> {
  fn clone(&self) -> Self {
    Self(PhantomData)
  }
}
```
2. Indeed, something we still have to support, should be tracked at #2. We still have to figure out how the syntax will look like then.

File: src/lib.rs (hunk 1 above)
Comment (daxpedda) [reply src/lib.rs]: Done.

File: src/lib.rs (hunk 2 above)
Comment (daxpedda) [reply src/lib.rs]: I did a little bit of research, using square brackets for the link should point to a reference. Like this:
```markdown
[link text][link reference]

[link reference]: https://example.com
```
But somehow, it recognizes that the reference name is a link and just uses it, I guess there is some built-in leniency in the markdown parser. I didn't dig into Rusts markdown rules.

File: src/lib.rs (hunk 3 above)
Comment (ModProg) [reply src/lib.rs:532]: Well probably just use empty bounds, we could use `look_ahead` and try parse the DeriveWhere and if it fails parse only Trait list.
so `#[derive_where(Clone)]` without `;` and `T`.

**Review by ModProg — CHANGES_REQUESTED** [review]
`build_struct` vs `generate_tuple`

document that the pattern parameter is the variant of an enum or Self we implement for

File: src/lib.rs (hunk 3 above)
Comment (ModProg) [reply src/lib.rs:532]: This is part of #2 and should not block this.

**Comment by daxpedda** [comment]
> `build_struct` vs `generate_tuple`

Don't know how I forgot that, I renamed them all to `build_for_x` now, I hope that's alright?
> document that the pattern parameter is the variant of an enum or Self we implement for

Right, you pointed that out above but I just forgot, sry 😅! I documented it only in `build_for_struct` and put a note in the others to look there.

**Review by ModProg — APPROVED** [review]

**Comment by ModProg** [comment]
LGTM, anything you want to add before merging?

**Comment by daxpedda** [comment]
No, looks good, I will start working on tests in a separate PR.

**Pull request merged**